Fails to build on ruby 3.3 with clang-1500.3.9.4

Given the last command line tools update on MacOS with a M1 iconv no longer builds.

Excerpt from the log:

```
iconv.c:709:50: error: incompatible function pointer types passing 'VALUE (VALUE, VALUE)' (aka 'unsigned long (unsigned long, unsigned long)') to parameter of type
'rb_block_call_func_t' (aka 'unsigned long (*)(unsigned long, unsigned long, int, const unsigned long *, unsigned long)') [-Wincompatible-function-pointer-types]
        rb_block_call(options, rb_intern("each"), 0, 0, get_iconv_opt_i, (VALUE)opt);
```

It does compile with `gem install iconv -- --with-cflags="-Wno-incompatible-function-pointer-types"`

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -14,7 +14,14 @@ jobs:
     strategy:
       matrix:
         ruby: ${{ fromJson(needs.ruby-versions.outputs.versions) }}
-        os: [ ubuntu-latest, macos-13 ]
+        os: [ ubuntu-latest, macos-13, macos-latest ]
+        exclude:
+          - ruby: 2.3
+            os: macos-latest
+          - ruby: 2.4
+            os: macos-latest
+          - ruby: 2.5
+            os: macos-latest
     runs-on: ${{ matrix.os }}
     steps:
     - uses: actions/checkout@v4

--- a/ext/iconv/iconv.c
+++ b/ext/iconv/iconv.c
@@ -188,7 +188,7 @@ static VALUE iconv_convert _((iconv_t cd, VALUE str, long start, long length, in
 static VALUE iconv_s_allocate _((VALUE klass));
 static VALUE iconv_initialize _((int argc, VALUE *argv, VALUE self));
 static VALUE iconv_s_open _((int argc, VALUE *argv, VALUE self));
-static VALUE iconv_s_convert _((struct iconv_env_t* env));
+static VALUE iconv_s_convert _((VALUE self));
 static VALUE iconv_s_iconv _((int argc, VALUE *argv, VALUE self));
 static VALUE iconv_init_state _((VALUE cd));
 static VALUE iconv_finish _((VALUE self));
@@ -204,7 +204,7 @@ static VALUE charset_map;
  * Returns the map from canonical name to system dependent name.
  */
 static VALUE
-charset_map_get(void)
+charset_map_get(VALUE klass)
 {
     return charset_map;
 }
@@ -642,7 +642,7 @@ iconv_s_allocate(VALUE klass)
 }
 
 static VALUE
-get_iconv_opt_i(VALUE i, VALUE arg)
+get_iconv_opt_i(RB_BLOCK_CALL_FUNC_ARGLIST(i, arg))
 {
     VALUE name;
 #if defined ICONV_SET_TRANSLITERATE || defined ICONV_SET_DISCARD_ILSEQ
@@ -784,8 +784,9 @@ iconv_s_open(int argc, VALUE *argv, VALUE self)
 }
 
 static VALUE
-iconv_s_convert(struct iconv_env_t* env)
+iconv_s_convert(VALUE self)
 {
+    struct iconv_env_t* env = (struct iconv_env_t*)self;
     VALUE last = 0;
 
     for (; env->argc > 0; --env->argc, ++env->argv) {
@@ -906,7 +907,7 @@ list_iconv(unsigned int namescount, const char *const *names, void *data)
 
 #if defined(HAVE_ICONVLIST) || defined(HAVE___ICONV_FREE_LIST)
 static VALUE
-iconv_s_list(void)
+iconv_s_list(VALUE klass)
 {
 #ifdef HAVE_ICONVLIST
     int state;

